mm2src/coins/utxo/utxo_tests.rs

```diff
@@ -4265,3 +4264,160 @@ fn test_utxo_validate_valid_and_invalid_pubkey() {
     assert!(coin.validate_other_pubkey(&[1u8; 20]).is_err());
     assert!(coin.validate_other_pubkey(&[1u8; 8]).is_err());
 }
+
+#[test]
+fn test_block_header_utxo_loop() {
+    static mut CURRENT_BLOCK_COUNT: u64 = 2048;
+
+    ElectrumClient::get_block_count
+        .mock_safe(|_| MockResult::Return(Box::new(futures01::future::ok(unsafe { CURRENT_BLOCK_COUNT }))));
+
+    let block_header_storage = BlockHeaderStorageForTests::new(TEST_COIN_NAME.to_string());
+
+    let block_header_storage_copy = block_header_storage.clone();
+    BlockHeaderStorage::new_from_ctx.mock_safe(move |_, _| {
+        MockResult::Return(Ok(BlockHeaderStorage {
+            inner: Box::new(block_header_storage_copy.clone()),
+        }))
+    });
+
+    let block_header_storage_copy = block_header_storage.clone();
+    BlockHeaderStorage::add_block_headers_to_storage.mock_safe(move |_, headers| {
+        let mut inner = block_on(block_header_storage_copy.inner.lock());
+        inner
+            .block_headers_by_hash
+            .extend(headers.iter().map(|(height, header)| (header.hash(), *height)));
+        inner.block_headers.extend(headers.into_iter());
+        MockResult::Return(Box::pin(futures::future::ok(())))
+    });
+
+    let conf = json!({"coin":"RICK","asset":"RICK","rpcport":8923,"enable_spv_proof": true});
+    let req = json!({
+         "method": "electrum",
+         "servers": [{"url":"electrum1.cipig.net:10017"}],
+    });
+    let ctx = MmCtxBuilder::new().into_mm_arc();
+    let params = UtxoActivationParams::from_legacy_req(&req).unwrap();
+    block_on(utxo_standard_coin_with_priv_key(
+        &ctx,
+        TEST_COIN_NAME,
+        &conf,
+        &params,
+        &[1u8; 32],
+    ))
+    .unwrap();
+    let block_headers_storage_mock = block_on(block_header_storage.inner.lock()).clone();
+
+    block_on(Timer::sleep(5.));
+    block_on(block_header_storage.add_block_headers_to_storage(generate_headers_map(2046..=2050))).unwrap();
+    // Check if there are blocks with 2046, 2047, 2048, 2049, 2050 heights.
+    for ((h1, _), h2) in block_headers_storage_mock.block_headers.iter().zip(2046_u64..=2050) {
+        assert_eq!(h1, &h2)
+    }
```

Comment:
This doesn't work as expected at the first glance.
[Iterator::zip](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.zip):
> If either iterator returns [None](https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.None), [next](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#tymethod.next) from the zipped iterator will return [None](https://doc.rust-lang.org/stable/std/option/enum.Option.html#variant.None).

I replaced this checking with the following, and it panics with
> Failed assertion None == Some(2046) for iteration 0
```rust
let actual = block_headers_storage_mock
        .block_headers
        .iter()
        .map(|(height, _header)| *height);
    itertools::assert_equal(actual, 2046_u64..=2050);
```

Please figure out what is wrong